src/docs/arch-s3-extension.dox

```diff
@@ -0,0 +1,79 @@
+
+S3 storage source, or the S3 store, is a WiredTiger extension that implements the storage source
+abstraction for AWS's S3 object-store. The store extends the WiredTiger's capability to store the
+data files on the S3 cloud. The extension provides a means to construct, configure and control a
+custom file system linked to an S3 bucket. The extension internally uses the API provided by the
+AWS C++ SDK to interact with the S3 cloud.
+
+The S3 store implements the following key functionalities defined by the storage source interface:
+
+Initialize the store, the AWS SDK and create a logging module for the store and the SDK.
+
+Register the S3 store as a storage source available to Wiredtiger.
+
+Provide an API to access and terminate the store. A reference count is incremented each time the
+storage source reference is taken and decremented with the termination. When the reference count
+goes to zero, the storage source is de-registered with WiredTiger and any allocated resources
+deallocated.
+
+Another essential functionality the S3 store provides is to create a WT_FILE_SYSTEM implementation
+for a given bucket. The store requires a bucket and an access token to create the filesystem. The
+bucket is expected to be a string argument, which is the bucket name and the corresponding region
+separated by a ';', e.g. "bucket1;ap-southeast-2". The AWS requires an access token to authenticate
+programmatic access to S3. The access token comprises an access key ID and a secret access key as a
+set. The S3 store requires the token to be provided as a single string argument of the access key
+followed by the secret key, separated by a ';'.
+
+The S3 store also configures a directory on the file system local to the WiredTiger database as a
+cache for the object files. The cache is meant to improve the read latency for the object files
+read after an initial download from the S3.
+
+@section s3_extension_connection S3 Connection class
+
+The S3Connection class represents an active connection to an S3 bucket. It encapsulates and
+implements all the operations the S3 filesystem requires. The instantiation of the class object
+authenticates with the AWS and validates the existence of and access to the configured bucket.
+The S3Connection exposes an API to list the bucket contents filtered by a directory and a prefix,
+check for an object's existence in the bucket, put an object to the cloud, and get the object from
+the cloud. Though not required for the file system's implementation, the class also provides the
+means to delete the objects to clean up artefacts from the internal unit testing.
+
+@section s3_extension_file_system S3_FILE_SYSTEM
+
+WiredTiger supports a database over a custom file system by giving a WT_FILE_SYSTEM interface for
+the end-user to implement. The S3 store itself provides a means to configure as many file systems
+as needed, each linked to an S3 bucket. These file systems, called the S3_FILE_SYSTEM, are a custom
```

Comment:
ditto `\c` in front of `S3_FILE_SYSTEM`

Reply:
Since there is no links generated for `S3_FILE_SYSTEM`, the `\c` could be great.